File: 产品资料/供应商资料/dobot_hex_sdk-main/realsense/src/camera_fps.cpp
/**
 * This file is Dobot Robotics's property. It contains Dobot Robotics trade secret, proprietary and confidential information.
 *
 * The information and code contained in this file is only for authorized Dobot Robotics employees to design, create, modify, or review.
 *
 * DO NOT DISTRIBUTE, DO NOT DUPLICATE OR TRANSMIT IN ANY FORM WITHOUT PROPER AUTHORIZATION.
 *
 * If you are not an intended recipient of this file, you must not copy, distribute, modify, or take any action in reliance on it.
 *
 * If you have received this file in error, please immediately notify Dobot Robotics and permanently delete the original and any copy of any file and any printout thereof.
 *
 * Copyright (c) 2025 Dobot Robotics
 *
 *
 * Author: Yixuan Chen
 * Mail: chenyixuan@dobot-robots.com
 * Created on Fri Sep 26 2025
 */

/**
 * 1. file function
 * @file camera_fps.cpp
 * @author  Yixuan Chen  (chenyixuan@dobot-robots.com)
 * @brief Subscribe Realsense depth and BGR images, calculate FPS and print basic info.
 * @version 0.1
 * @date 2025-09-26
 *
 * @copyright Copyright (c) 2025
 *
 */

/**
 * 2.C/C++ basic header files declaration
 */
#include <algorithm>
#include <cstdint>
#include <cstring>
#include <limits>
#include <memory>
#include <string>
#include <fstream>
#include <filesystem>
#include <iostream>

/**
 * 3.ROS2 module headers declaration
 */
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>

#define PROGRAM_MAJOR_VERSION 1
#define PROGRAM_MINOR_VERSION 0

/**
 * @class DepthLiteCameraFPSSubscriber
 * @brief ROS2 node for subscribing depth and BGR image topics, calculating FPS and printing info.
 *
 * This class subscribes to Realsense depth (mono16) and BGR image topics.
 * It computes instantaneous FPS and smoothed FPS (EMA) and prints them periodically.
 */
class DepthLiteCameraFPSSubscriber : public rclcpp::Node
{
public:
    /**
     * @brief Construct a new Depth Lite Subscriber object
     *
     * @param strNodeName
     * @version 0.1
     * @author  Yixuan Chen
     * @date 2025-09-26
     * @copyright Copyright (c) 2025
     */
    DepthLiteCameraFPSSubscriber(const std::string &strNodeName) : Node(strNodeName)
    {
        using std::placeholders::_1;

        // subscriber
        auto qos = rclcpp::SensorDataQoS();
        m_pDepthImageSubscriber = this->create_subscription<sensor_msgs::msg::Image>(
            m_strDepthImageTopicName, qos, std::bind(&DepthLiteCameraFPSSubscriber::depthCallback, this, _1));
        m_pBGRImageSubscriber = this->create_subscription<sensor_msgs::msg::Image>(
            m_strBGRImageTopicName, qos, std::bind(&DepthLiteCameraFPSSubscriber::bgrCallback, this, _1));

        // fps init
        m_lastStamp_depth = this->now();
        m_lastStamp_bgr = this->now();
        m_emaFps_depth = 0.0;
        m_emaFps_bgr = 0.0;
    }

private:
    void depthCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg);
    void bgrCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg);

private:
    // topic names
    std::string m_strDepthImageTopicName = "/realsense_camera_node/sn408122070053/depth/u16/image_raw";
    std::string m_strBGRImageTopicName = "/realsense_camera_node/sn408122070053/color/bgr/image_raw";

    // subscription
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr m_pDepthImageSubscriber = nullptr;
    rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr m_pBGRImageSubscriber = nullptr;

    // fps
    rclcpp::Time m_lastStamp_depth;
    rclcpp::Time m_lastStamp_bgr;
    double m_emaFps_depth;
    double m_emaFps_bgr;

    int cnt = 0; // counter for alternating log outputs
};

/**
 * @brief Depth image callback, calculate FPS and print info
 *
 * @param msg
 * @version 0.1
 * @author  Yixuan Chen
 * @date 2025-09-26
 * @copyright Copyright (c) 2025
 */
void DepthLiteCameraFPSSubscriber::depthCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg)
{
    // FPS calculation
    const auto now = this->now();
    const double dt = (now - m_lastStamp_depth).seconds();
    m_lastStamp_depth = now;

    if (dt > 1e-6)
    {
        const double inst_fps = 1.0 / dt;
        const double alpha = 0.2; // EMA 平滑
        m_emaFps_depth = (m_emaFps_depth == 0.0) ? inst_fps
                                                 : (alpha * inst_fps + (1.0 - alpha) * m_emaFps_depth);
    }
    // log information
    cnt += 1;
    if (cnt % 2 == 1)
    {
        RCLCPP_INFO(this->get_logger(),
                    "FPS: %.2f (inst=%.2f) | enc=%s WxH=%ux%u",
                    m_emaFps_depth,
                    (dt > 1e-6 ? 1.0 / dt : 0.0),
                    msg->encoding.c_str(), msg->width, msg->height);
    }
}

/**
 * @brief BGR image callback, calculate FPS and print info
 *
 * @param msg
 * @version 0.1
 * @author  Yixuan Chen
 * @date 2025-09-26
 * @copyright Copyright (c) 2025
 */
void DepthLiteCameraFPSSubscriber::bgrCallback(const sensor_msgs::msg::Image::ConstSharedPtr &msg)
{
    // FPS calculation
    const auto now = this->now();
    const double dt = (now - m_lastStamp_bgr).seconds();
    m_lastStamp_bgr = now;

    if (dt > 1e-6)
    {
        const double inst_fps = 1.0 / dt;
        const double alpha = 0.2; // EMA smooth factor
        m_emaFps_bgr = (m_emaFps_bgr == 0.0) ? inst_fps
                                             : (alpha * inst_fps + (1.0 - alpha) * m_emaFps_bgr);
    }
    // log information
    cnt += 1;
    if (cnt % 2 == 0)
    {
        RCLCPP_INFO(this->get_logger(),
                    "FPS: %.2f (inst=%.2f) | enc=%s WxH=%ux%u",
                    m_emaFps_bgr,
                    (dt > 1e-6 ? 1.0 / dt : 0.0),
                    msg->encoding.c_str(), msg->width, msg->height);
    }
}

int main(int argc, char **argv)
{
    if (argc > 1 && std::string(argv[1]) == "-v")
    {
        std::string version = std::to_string(PROGRAM_MAJOR_VERSION) + "." + std::to_string(PROGRAM_MINOR_VERSION);
        std::cout << version << std::endl;
        return 0;
    }

    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<DepthLiteCameraFPSSubscriber>("camera_fps_subscriber"));
    rclcpp::shutdown();
    return 0;
}

File: 产品资料/供应商资料/dobot_hex_sdk-main/realsense/py/camera_fps.py
"""
This file is Dobot Robotics's property. It contains Dobot Robotics trade secret, proprietary and confidential information.

The information and code contained in this file is only for authorized Dobot Robotics employees to design, create, modify, or review.

DO NOT DISTRIBUTE, DO NOT DUPLICATE OR TRANSMIT IN ANY FORM WITHOUT PROPER AUTHORIZATION.

If you are not an intended recipient of this file, you must not copy, distribute, modify, or take any action in reliance on it.

If you have received this file in error, please immediately notify Dobot Robotics and permanently delete the original and any copy of any file and any printout thereof.

Copyright (c) 2025 Dobot Robotics


Author: Yixuan Chen
Mail: chenyixuan@dobot-robots.com
Created on Fri Sep 26 2025
"""

"""
1. file function
@file camera_fps.py
@author  Yixuan Chen
@brief Subscribe Realsense depth (mono16) and BGR image topics, calculate FPS, and print info.
       (NOTE: This Python port prints FPS like the C++ version; PGM saving stub left for extension.)
@version 0.1
@date 2025-09-26
"""
import sys
from typing import Optional

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image


PROGRAM_MAJOR_VERSION = 1
PROGRAM_MINOR_VERSION = 0


class DepthLiteCameraFPSSubscriber(Node):
    """
    ROS2 node for subscribing to Realsense depth and BGR images.

    This node:
      - Subscribes to mono16 depth and BGR image topics
      - Computes instantaneous and smoothed FPS (EMA)
      - Logs FPS and image encoding/size periodically
    """

    def __init__(self, node_name: str = "camera_fps_subscriber"):
        """
        Initialize the DepthLiteCameraFPSSubscriber node.

        @param node_name (str): name of the ROS2 node
        """
        super().__init__(node_name)

        # params
        self.depth_topic = "/realsense_camera_node/sn408122070053/depth/u16/image_raw"
        self.bgr_topic = "/realsense_camera_node/sn408122070053/color/bgr/image_raw"

        # fps state
        now = self.get_clock().now()
        self._last_depth_stamp = now
        self._last_bgr_stamp = now
        self._ema_fps_depth: float = 0.0
        self._ema_fps_bgr: float = 0.0
        self._cnt: int = 0

        qos = rclpy.qos.qos_profile_sensor_data

        # subscribers
        self._sub_depth = self.create_subscription(
            Image, self.depth_topic, self.depth_callback, qos
        )
        self._sub_bgr = self.create_subscription(
            Image, self.bgr_topic, self.bgr_callback, qos
        )

        self.get_logger().info(
            f"Subscribers started. depth='{self.depth_topic}', bgr='{self.bgr_topic}'"
        )

    @staticmethod
    def _secs_between(now, last) -> float:
        """
        Compute seconds between two ROS2 time stamps.

        @param now (rclpy.time.Time): current timestamp
        @param last (rclpy.time.Time): previous timestamp
        @return float: time difference in seconds
        """
        return (now - last).nanoseconds / 1e9

    @staticmethod
    def _ema(prev: float, x: float, alpha: float = 0.2) -> float:
        """
        Exponential moving average.

        @param prev (float): previous EMA value
        @param x (float): new input value
        @param alpha (float): smoothing factor
        @return float: updated EMA value
        """
        return x if prev == 0.0 else (alpha * x + (1.0 - alpha) * prev)

    def depth_callback(self, msg: Image):
        """
        Callback for depth image topic.

        Calculates instantaneous and smoothed FPS for depth stream,
        and logs results every other callback.

        @param msg (Image): ROS2 Image message
        """
        now = self.get_clock().now()
        dt = self._secs_between(now, self._last_depth_stamp)
        self._last_depth_stamp = now

        if dt > 1e-6:
            inst = 1.0 / dt
            self._ema_fps_depth = self._ema(self._ema_fps_depth, inst, 0.2)

        self._cnt += 1
        if self._cnt % 2 == 1:
            inst_show = (1.0 / dt) if dt > 1e-6 else 0.0
            self.get_logger().info(
                f"FPS: {self._ema_fps_depth:.2f} (inst={inst_show:.2f}) | "
                f"enc={msg.encoding} WxH={msg.width}x{msg.height}"
            )

    def bgr_callback(self, msg: Image):
        """
        Callback for BGR image topic.

        Calculates instantaneous and smoothed FPS for BGR stream,
        and logs results every other callback.

        @param msg (Image): ROS2 Image message
        """
        now = self.get_clock().now()
        dt = self._secs_between(now, self._last_bgr_stamp)
        self._last_bgr_stamp = now

        if dt > 1e-6:
            inst = 1.0 / dt
            self._ema_fps_bgr = self._ema(self._ema_fps_bgr, inst, 0.2)

        self._cnt += 1
        if self._cnt % 2 == 0:
            inst_show = (1.0 / dt) if dt > 1e-6 else 0.0
            self.get_logger().info(
                f"FPS: {self._ema_fps_bgr:.2f} (inst={inst_show:.2f}) | "
                f"enc={msg.encoding} WxH={msg.width}x{msg.height}"
            )


def main(args=None):
    # 版本输出
    if len(sys.argv) > 1 and sys.argv[1] == "-v":
        print(f"{PROGRAM_MAJOR_VERSION}.{PROGRAM_MINOR_VERSION}")
        return

    rclpy.init(args=args)
    node = DepthLiteCameraFPSSubscriber()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()
